src/qs8-gemm/MRx16c2s2-wasmsimd-dot16x2.c.in

```diff
@@ -0,0 +1,337 @@
+  assert(mr <= ${MR});
+  assert(nc != 0);
+  assert(kc != 0);
+  assert(kc % sizeof(int8_t) == 0);
+  assert(a != NULL);
+  assert(w != NULL);
+  assert(c != NULL);
+
+  kc = round_up_po2(kc, 4 * sizeof(int8_t));
+  const int8_t* a0 = a;
+  ${OUT_T}* c0 = c;
+  $for M in range(1, MR):
+    const int8_t* a${M} = (const int8_t*) ((uintptr_t) a${M-1} + a_stride);
+    ${OUT_T}* c${M} = (${OUT_T}*) ((uintptr_t) c${M-1} + cm_stride);
+    $if M % 2 == 0:
+      if XNN_UNPREDICTABLE(mr <= ${M}) {
+        a${M} = a${M-1};
+        c${M} = c${M-1};
+      }
+    $elif M + 1 == MR:
+      if XNN_UNPREDICTABLE(mr != ${M+1}) {
+        a${M} = a${M-1};
+        c${M} = c${M-1};
+      }
+    $else:
+      if XNN_UNPREDICTABLE(mr < ${M+1}) {
+        a${M} = a${M-1};
+        c${M} = c${M-1};
+      }
+
+  $if DATATYPE == "QD8":
+    const v128_t vmin = wasm_v128_load32_splat(&params->scalar.min);
+    const v128_t vmax = wasm_v128_load32_splat(&params->scalar.max);
+    XNN_FORCE_REALIZATION(vmin);
+    XNN_FORCE_REALIZATION(vmax);
+  $else:
+    $if DATATYPE != "QC8":
+      const v128_t vscale = wasm_v128_load32_splat(&params->fp32_scalar.scale);
+      XNN_FORCE_REALIZATION(vscale);
+
+    const v128_t vmagic_bias = wasm_f32x4_const_splat(12582912.0f);
+    const int32_t output_min_less_zero_point = (int32_t) params->fp32_scalar.output_min - (int32_t) params->fp32_scalar.output_zero_point;
+    const v128_t vmagic_min = wasm_i32x4_splat((int32_t) float_as_uint32(12582912.0f + output_min_less_zero_point));
+    const v128_t vmagic_bias_less_output_zero_point = wasm_i32x4_splat(INT32_C(0x4B400000) - (int32_t) params->fp32_scalar.output_zero_point);
+    const v128_t voutput_max = wasm_i8x16_splat(params->fp32_scalar.output_max);
+    //XNN_FORCE_REALIZATION(vmagic_bias);
+    //XNN_FORCE_REALIZATION(vmagic_min);
+    //XNN_FORCE_REALIZATION(vmagic_bias_less_output_zero_point);
+    //XNN_FORCE_REALIZATION(voutput_max);
+
+  do {
+    $if DATATYPE == "QD8":
+      v128_t vksum${ABC[0:4]} = wasm_v128_load((const int32_t*) w);
+      $for N in range(4, NR, 4):
+        v128_t vksum${ABC[N:N+4]} = wasm_v128_load((const int32_t*) w + ${N});
+      $for M in range(MR):
+        const v128_t vinput_zero_point${M} = wasm_v128_load32_splat(&quantization_params[${M}].zero_point);
+        $for N in range(0, NR, 4):
+          v128_t ${VACC(M)}x${ABC[N:N+4]} = wasm_i32x4_mul(vksum${ABC[N:N+4]}, vinput_zero_point${M});
+    $else:
+      v128_t ${VACC(0)}x0123 = wasm_v128_load(w);
+      $for N in range(4, NR, 4):
+        v128_t ${VACC(0)}x${ABC[N:N+4]} = wasm_v128_load((const int32_t*) w + ${N});
+      $for M in range(1, MR, 1):
+        $for N in range(0, NR, 4):
+          v128_t ${VACC(M)}x${ABC[N:N+4]}= ${VACC(0)}x${ABC[N:N+4]};
+    w = (const int32_t*) w + ${NR};
+
+    size_t k = kc;
+    $if UNROLL > 1:
+      while (k >= ${UNROLL * 4} * sizeof(int8_t)) {
+        $for M in range(MR):
+          v128_t va${M}x0x0123 = wasm_v128_load32_splat(a${M});
```

Comment:
instead of s2 with 32 bit splat, consider c2s4 - a 64 bit load (8 bytes) extended to 8 shorts.
instead of  2 dot products (of c2) with 1 shuffle after the dot product, unroll to do 4 dot products with 3 shuffles.  no additional registers needed
with s2 processing 64 bits (8 channels)
2 broadcast of 32 bits
2 sign extend to 16 bits
4 dot product
2 shuffles
with s4
1 sign extend load
4 dot product
3 shuffles